examples/gno.land/p/demo/paginate/paginate.gno

```diff
@@ -9,25 +16,84 @@ type renderer interface {
 	Render() string
 }
 
-// TODO this function belongs somewhere convenient.
-// escape a URL for markdown.
-func escMDURL(s string) string {
-	// XXX implement
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
 }
 
-var itoa = strconv.Itoa
+func navigation(numPages int, current int, reverse bool, pageFn pageFn) string {
+	str := ""
+
+	i := 1
+	upperLimit := min(PaginateGap+1, current)
+	for ; i < upperLimit; i++ {
+		pageurl := pageFn(i, numPages, reverse)
+		str += md.Link(strconv.Itoa(i), pageurl) + " "
+	}
+	if i-1 == PaginateGap && i != current {
+		if i == current-PaginateGap/2 {
+			pageurl := pageFn(i, numPages, reverse)
+			str += md.Link(strconv.Itoa(i), pageurl) + " "
+			i += 1
+		} else {
+			str += "... "
+		}
+	}
+
+	// sanity check
+	if current > numPages {
+		for i = numPages - PaginateGap + 1; i <= numPages; i++ {
+			pageurl := pageFn(i, numPages, reverse)
+			str += md.Link(strconv.Itoa(i), pageurl) + " "
+		}
+		return str
+	}
+
+	if i < current-PaginateGap/2 {
+		i = current - PaginateGap/2
+	}
+
+	upperLimit = min(current+PaginateGap/2+1, numPages)
+	for ; i < upperLimit; i++ {
+		if i == current {
+			str += "*" + strconv.Itoa(i) + "* "
+			continue
+		}
+		pageurl := pageFn(i, numPages, reverse)
+		str += md.Link(strconv.Itoa(i), pageurl) + " "
+	}
+	if i < numPages-PaginateGap+1 {
+		if i == numPages-PaginateGap {
+			pageurl := pageFn(i, numPages, reverse)
+			str += md.Link(strconv.Itoa(i), pageurl) + " "
+		} else {
+			str += "... "
+		}
+		i = numPages - PaginateGap + 1
+	}
+	for ; i <= numPages; i++ {
+		pageurl := pageFn(i, numPages, reverse)
+		str += md.Link(strconv.Itoa(i), pageurl) + " "
+	}
+	return str
+}
+
+// This syntax actually doesn't work with Gno language
+// var itoa = strconv.Itoa
 
 // pageFn: path function to render paginattion links. if nil, no links are
 // rendered. result must start with ':'.
 // t: the tree with elements. each value must implement Render() string.
 // page: page number starting from 1.
 // size: page size.
-// reverse: reverse order. // XXX implement
+// reverse: reverse order.
 func RenderTreeByPage(
```

Comment:
What about exposing a `p/demo/ui.Pager` struct instead ? (cc @moul )

Reply:
if generic yes, else, we can make UI extendable with composition. 👍

Reply:
Keep pagination separate from `ui`. Consider:

* `ui`: A minimal, independent UI for future composition.
* `paginate`: A standalone pagination engine.
* `gnoostrap`: An opinionated collection utilizing `ui`, `paginate` and other packages, akin to Twitter's Bootstrap.